test_launch_ros/test/test_launch_ros/test_normalize_parameters.py

```diff
@@ -0,0 +1,173 @@
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_multiple_dictionaries():
+    orig = [{'foo': 1, 'bar': 2.0}, {'baz': 'asdf'}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 1, 'bar': 2.0}, {'baz': 'asdf'})
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_substitution():
+    orig = [{TextSubstitution(text='bar'): TextSubstitution(text='baz')}]
+    norm = normalize_parameters(orig)
+    expected = ({'bar': 'baz'},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_substitution_list_name():
+    orig = [{(TextSubstitution(text='bar'), TextSubstitution(text='foo')): 1}]
+    norm = normalize_parameters(orig)
+    expected = ({'barfoo': 1},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_substitution_list_value():
+    orig = [{'foo': [TextSubstitution(text='fiz'), TextSubstitution(text='buz')]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 'fizbuz'},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+    orig = [{'foo': [[TextSubstitution(text='fiz')], [TextSubstitution(text='buz')]]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': ('fiz', 'buz')},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_mixed_substitutions_and_strings():
+    orig = [{'foo': [TextSubstitution(text='fiz'), 'bar']}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 'fizbar'},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+    orig = [{'foo': [[TextSubstitution(text='fiz')], 'bar']}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': ('fiz', 'bar')},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_str():
+    orig = [{'foo': 'bar', 'fiz': ['b', 'u', 'z']}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 'bar', 'fiz': ('b', 'u', 'z')},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_bool():
+    orig = [{'foo': False, 'fiz': [True, False, True]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': False, 'fiz': (True, False, True)},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_float():
+    orig = [{'foo': 1.2, 'fiz': [2.3, 3.4, 4.5]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 1.2, 'fiz': (2.3, 3.4, 4.5)},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_int():
+    orig = [{'foo': 1, 'fiz': [2, 3, 4]}]
+    norm = normalize_parameters(orig)
+    expected = ({'foo': 1, 'fiz': (2, 3, 4)},)
+    assert evaluate_parameters(LaunchContext(), norm) == expected
+
+
+def test_dictionary_with_int_and_float():
+    orig = [{'foo': 1, 'fiz': [2, 3.1, 4]}]
```

Comment:
With knowledge of the implementation, I think it wouldn't be a bad idea to add a test case where the `float` is the first element in the list.

Reply:
b1f22861830e10212d8176e9a069226d78077400